src/Context/Traits/UserAwareContextTrait.php

```diff
@@ -84,31 +85,134 @@ public function loggedIn()
+     * throws an UnexpectedValueException if not
+     *
+     * if there is return a user ID
+     *
+     * @throws \UnexpectedValueException
+     * @return \WP_User $user
+     */
+    private function getExistingMatchingUser($args)
+    {
+        $user_id = $this->getUserIdFromLogin($args['user_login']);
+        $user = $this->getDriver()->user->get($user_id);
+
+        /* users can have more than one role so needs to be a special case */
+        if (array_key_exists('role', $args)) {
+            $this->checkUserHasRole($user, $args['role']);
+        }
+
+        /*
+         * Loop through each of the passed arguements.
+         * if they are arguments which apply to users
+         * then check that that which exist matches that which was specified.
+         */
+        foreach ($args as $parameter => $value) {
+            if ($parameter == 'password') {
+                try {
+                    if (! $this->getDriver()->user->validateCredentials($args['user_login'], $value)) {
+                        throw new \UnexpectedValueException('User with login : ' . $user->user_login . ' exists but password is incorrect');
+                    }
+                } catch (UnsupportedDriverActionException $exception) {
+                    // WPCLI can't do this yet.
+                }
+            }
+            if ($this->isValidUserParameter($parameter) && $user->$parameter != $args[$parameter]) {
+                throw new \UnexpectedValueException('User with login : ' . $user->user_login . 'exists, but ' . $parameter . ' is ' . $user->$parameter . ' not ' . $args[$parameter] . 'which was specified');
+            }
+        }
+
+        return $user;
+    }
+
+    /**
+     * Checks to see if the user has an assigned role or not.
+     *
+     * @param \WP_User $user
+     * @param string $role
+     *
+     * @throws \UnexpectedValueException
+     *
+     * @return boolean $retval True if the role does apply to the user.
+     */
+    private function checkUserHasRole($user, $role)
+    {
+        // $user->role can either be a string with 1 role in it or an array of roles.
+        if (is_string($user->roles)) {
+            if ($user->roles != $role) { // if its a string check it matches the role
+                throw new \UnexpectedValueException('User with login : ' . $user->user_login . ' exists, but role : ' . $role . ' does not match the applied role : ' . $user->roles);
+            }
+        } elseif (! in_array($role, $user->roles)) { // if its an array check the role is in that array
+            throw new \UnexpectedValueException('User with login : ' . $user->user_login . ' exists, but role : ' . $role . ' is not in the list of applied roles : ' . $user->roles);
+        }
+
+        return true;
+    }
+
+    /**
+     * Checks to see if the passed in parameter applies to a user or not.
+     *
+     * @param string $user_parameter
+     *            the parameter to be checked.
```

Comment:
One line for this phpdoc please

Reply:
Yeah - that's my IDE being rubbish - i've fixed it now.